### DISCUSSION TITLE
Starting Point, CLI

### DISCUSSION THREAD

**Discussion by Sarctiann** [kkinder/puepy#57] [Q&A]
## The problem

Hello, I'm super new here. I was taking a look at this project and it looks GREAT! I'd love to contribute, or at least use it 🥲.

I'd appreciate a `fastapi` integration starting point. I tried to install and run `puepy` via `poetry` then created the [`index.html`, `main.py`, `pyscript.json`] and downloaded the *wheel* file...

But I didn't figure out how to serve the "minimal app" I tried to run just `python3 main.py` but of course, this didn't work, (I can imagine that file should be sent to the client).

So I cloned the repository and used the `serve_examples.py` which worked as expected, but I don't know how to decouple that "not than minimal" examples project into an app's starting point.

---

## The Proposal

In Summary, I wonder if you could provide a minimal CLI (i.e. `pue`) with at least three commands `init or new`, `dev or run`, and `build` (in the future this could grow to include common backend integrations like Django, Flask, or FastApi ❤️ )

### `pue init` or `pue new`

So, like in `vue`, `react`, `angular`, or almost any fe framework, we can start a minimal app with the basic files str/(`index.html`, `__init__.py`, `styles.css`) **and please** the required files to run (`pyscript.json`, the **wheel file**, etc.)

### `pue dev` or `pue run`

starts a simpler (or not) live server to start working on the app

### `pue build`

To put everything together into a `templates/` or `static/` folder to be served with Django, flask, FastAPI, or any of that kind...

---

Thanks for doing this!

**Comment by Sarctiann** [comment]
Hey! I got a working example with Fastapi

![image](https://github.com/user-attachments/assets/3dacacd4-eab9-416e-860f-86034a8d1b8d)

Basic for now...

**Comment by kkinder** [comment]
I originally had in mind that framework integration would be via separate repos. For example, [I made a little experiment for Django](https://github.com/kkinder/django-puepy/). One thing that's different from Vue/React/Angular is that with PuePy, there's no build step at all, so commands like `pue build` and `pue dev` don't make as much sense; all `pue dev` would do, presumably, is run a local web server.

Are you familiar with [cookiecutter](https://github.com/cookiecutter/cookiecutter)? I think it would probably be the best way of unpacking starter projects.

BTW, you can run a Python web server just with this:

```
python3 -m http.server 8888
```

**Reply by Sarctiann** [reply]
Thank you! 
Honesty, after explaining "my problem" I realized that the Python app file should be sent to the client 🤣. so I went to my first-attempt files (index.html, main.py, ...puepy...whl, pyscript.json) and ran `python3 -m http.server` in that directory and it works! I'm still shocked (I Never used pyscript before)

I also noticed that there is no transpilation/compilation/building process at any point. 😱 Amazing.

I will investigate "cookiecutter", Thank you for your advice!

---

I took a look at `django-puepy` and I noticed it requires some "not pure" approach, I guess it might be the only/better way to do it... But I'd like to see a more "pure" approach. I mean an approach that establishes one way to write "PuePy" that allows you to switch the "backend" more easily. I think Jinja templates are absolutely fine since they are "the common way" to send statics from almost every python's backend framework/library/package. So I guess the minimal integrable app could be made with two folders "static" and "templates"

---

### In summary (new summary)

The ideal flow to me, is something like this:

```shell
~ $ pip install puepy

# ...

~ $ puepy new my_app 

# generates:
#    my_app/
#    |_ index.html
#    |_ my_app.py
#    |_ pyscript.json
#    |_ puepy-...-any.whl

# Then it would be nice to have an alias for `python3 -m http.server --whatever`

~ $ puepy serve

# or the integrable way

~ $ cd my_api_project

~/my_api_project $ puepy init
The name of the static directory [static]: _
The name of the templates directory [templates]: _

# generates:
#    #(my_api_project/)
#    |_ static/
#    |   |_ my_app.py
#    |   |_ pyscript.json
#    |   |_ puepy-...-any.whl
#    |_ templates/
#        |_ index.html

~/my_api_project $ python3 run_api.py
# Which of course requires the user to define proper routes... but we can live with that haha

# and if we can provide a tiny server-side jinja2 templating to `puepy serve` so we can run the front-end only
~/my_api_project $ puepy serve -j templates

```
